janus_server/src/datastore.rs

```diff
@@ -0,0 +1,318 @@
+use prio::codec::{decode_u16_items, encode_u16_items, CodecError, Decode};
+use tokio_postgres::{error::SqlState, IsolationLevel, Row};
+
+use crate::message::{Extension, HpkeCiphertext, Nonce, Report, TaskId, Time};
+use std::{future::Future, io::Cursor, pin::Pin};
+use tokio::sync::Mutex;
+
+// TODO(brandon): connection pooling (currently all concurrent transactions in a single Datastore are serialized)
+// TODO(brandon): prepare DB statements
+
+/// Datastore represents a datastore for Janus, with support for transactional reads and writes.
+/// In practice, Datastore instances are currently backed by a PostgreSQL database.
+pub struct Datastore {
+    db: Mutex<tokio_postgres::Client>,
+}
+
+impl Datastore {
+    /// new creates a new Datastore using the given Client for backing storage. It is assumed that
+    /// the Client is connected to a database with a compatible version of the Janus database schema.
+    pub fn new(db: tokio_postgres::Client) -> Datastore {
+        Self { db: Mutex::new(db) }
+    }
+
+    /// run_tx runs a transaction, whose body is determined by the given function. The transaction
+    /// is committed if the body returns a successful value, and rolled back if the body returns an
+    /// error value.
+    ///
+    /// The datastore will automatically retry some failures (e.g. serialization failures) by
+    /// rolling back & retrying with a new transaction, so the given function should support being
+    /// called multiple times. Values read from the transaction should not be considered as
+    /// "finalized" until the transaction is committed, i.e. after `run_tx` is run to completion.
+    pub async fn run_tx<F, T>(&self, f: F) -> Result<T, Error>
+    where
+        for<'a> F: Fn(&'a Transaction) -> Pin<Box<dyn Future<Output = Result<T, Error>> + 'a>>,
+    {
+        loop {
+            let rslt = self.run_tx_once(&f).await;
+            if let Some(err) = rslt.as_ref().err() {
+                if err.is_serialization_failure() {
+                    continue;
+                }
```

Comment:
Besides the case of serialization failures, what's our story for retries in the face of network problems? Does `tokio-postgres` do something sensible with exponential backoffs and so on or are we on the hook for implementing that based on CloudSQL documentation?

Reply:
tokio_postgres doesn't document anything in terms of automatic retries, nor does it document what happens if it encounters a network failure. We might want to do something better here, but we'll need to see what the errors actually look like before we can discriminate based on them.

Due to the stateful nature of a DB connection (as opposed to cloud APIs which are typically stateless from request to request in a connection), I strongly suspect tokio_postgres doesn't e.g. automatically retry a failed `query` call. I think we'd want to notice a network error at the level of `run_tx`, drop the client/connection in use, and grab a new connection from the pool to try again.

Reply:
OK. We should write a TODO and/or an issue about retries then.